filecoin-proofs/src/api/mod.rs

```diff
@@ -1,409 +1,167 @@
+                    // println!(
+                    //     "S2 <- H(s({}) | {} ({}))",
+                    //     u64::from(stack[stack.len() - 1].size) / 127,
+                    //     i,
+                    //     u64::from(piece.size) / 127
+                    // );
+                    {
+                        let l = stack.len();
+                        let el = &mut stack[l - 1];
+                        el.size = UnpaddedBytesAmount(u64::from(el.size) + u64::from(piece.size));
+                        let h = piece_hash(&el.commitment, &piece.commitment);
+                        el.commitment.copy_from_slice(h.as_ref());
+                    }
+
+                    // try reduce stack
+                    while stack.len() > 1
+                        && stack[stack.len() - 2].size == stack[stack.len() - 1].size
+                    {
+                        let l = stack.len();
+                        // println!(
+                        //     "  S2[{}] <- H(S2[{}]({}) | S2[{}]({}))",
+                        //     l - 2,
+                        //     l - 2,
+                        //     u64::from(stack[l - 2].size) / 127,
+                        //     l - 1,
+                        //     u64::from(stack[l - 1].size) / 127
+                        // );
+
+                        stack[l - 2].size = UnpaddedBytesAmount(
+                            u64::from(stack[l - 2].size) + u64::from(stack[l - 1].size),
+                        );
+                        let h = piece_hash(&stack[l - 2].commitment, &stack[l - 1].commitment);
+                        stack[l - 2].commitment.copy_from_slice(h.as_ref());
+
+                        // remove last element from stack
+                        stack.pop();
+                    }
+
+                    // Reduce to the current if stack only contains one element.
+                    if stack.len() == 1 && stack[0].size == current.size {
+                        // println!(
+                        //     "  current <- H(current({}) | S[0]({}))",
+                        //     u64::from(current.size) / 127,
+                        //     u64::from(stack[0].size) / 127
+                        // );
+                        current.size =
+                            UnpaddedBytesAmount(u64::from(current.size) + u64::from(stack[0].size));
+                        let h = piece_hash(&current.commitment, &stack[0].commitment);
+                        current.commitment.copy_from_slice(h.as_ref());
+
+                        // remove el from stack
+                        stack.pop();
+                    }
+                } else {
+                    // println!("S2 <- {}({})", i, u64::from(piece.size) / 127);
+                    stack.push(piece.clone());
+                }
+            }
+            Ordering::Greater => {
+                unreachable!("algorithm failed");
+            }
+        }
```

Comment:
I think this still isn't right. I've listed one example of how above. I think it's easiest just to describe what I think the right algorithm is. I will also try to write this into the spec (in code) for the sake of clarity and lack of ambiguity. I think the quirks of trying to write this 'efficiently' and in Rust aren't helping us.

I think once we assume padding pieces have been added *correctly*, the algorithm i simply this:

- For each successive piece, push it onto the output stack and fully reduce the stack.
- The stack can be reduced whenever the top two pieces have the same size.
- Reduction involves popping both pieces off the stack and pushing the result of hashing them.
- 'Fully reducing' means examining the stack after each reduction and performing another reduction if necessary.

This is almost exactly the algorithm I implemented in my experimental 'streaming merkle trees' PR: https://github.com/filecoin-project/merkle_light/pull/26/files#diff-be3695d843cbc8410e25ba4e24f978a5R77-R143

The main difference here is that instead of always adding `Roots` with height 0, (corresponding to a piece of size 1 node), we can push roots representing subtrees with different heights onto the stack. It's still the case that a root with greater height than the top of the stack can never be pushed, but this requires that piece padding was added correctly.

If it were ever the case that the next piece to be added had a root height *greater* than that on top of the stack, it would mean that piece padding should have been added.

This observation suggests what might be the cleanest algorithm: instead of adding piece padding in its own step, instead apply the reduction algorithm (as described above). Whenever shifting the next piece would produce an illegal state (one violating the invariant that roots heights on the stack never increase — think 'towers of Hanoi'), instead push a commitment to padding of the same size as the piece whose root is on top of the stack.

This will allow adding exactly the right padding, in one pass, without dependencies on potentially more complicated ways of calculating it, and in a way which makes clear *why* the chosen padding piece size is correct.

I am going to prioritize putting this in the spec as the fastest way to clarify any remaining questions either of us might have.

Reply:
By the way, I think what you have now is probably close to this (ignoring the padding part). I wrote it out prescriptively since that was easier and less error-prone than trying to identify the (perhaps small) differences.

Reply:
I wrote this into the spec: https://github.com/filecoin-project/specs/pull/586

My intention was to illustrate what I described above as straightforwardly as possible. It's late and spec lacks tests, so it's possible something isn't quite right. But this should get you most of the way there.

Reply:
https://github.com/filecoin-project/specs/blob/b14a003a9bc144a77e7068f8cc018bc3ec72e278/src/libraries/filcrypto/filproofs/algorithms.go#L419-L482

Reply:
https://github.com/filecoin-project/specs/blob/b14a003a9bc144a77e7068f8cc018bc3ec72e278/src/libraries/filcrypto/filproofs/algorithms.go#L419-L482

Reply:
https://github.com/filecoin-project/specs/blob/de114990b925351bb86b6406c37e84ac1dceffcd/src/libraries/filcrypto/filproofs/algorithms.go#L419-L482

Reply:
all finally fixed